File: lib/solvers/HighDensitySolver/MultiHeadPolyLineIntraNodeSolver/getPossibleInitialViaPositions.ts
import type { Bounds, Point } from "@tscircuit/math-utils"
import { getCentroidsFromInnerBoxIntersections } from "./getCentroidsFromInnerBoxIntersections"
import { generateBinaryCombinations } from "./generateBinaryCombinations"
import { MHPoint } from "./types1"

type ViaPositionVariantForLinesViaCountVariant = {
  viaPositions: Point[]
  viaCountVariant: number[]
}

/**
 * Get the all possible via positions if you consider the centroids of shapes
 * created by all the intersections that divide the bounding box of the node
 *
 * We iterate over the via count variants, for each variant we generate each
 * possible via position.
 * Remember: The viaCountVariants specifies the number of vias for each connection,
 * so that's how we know how many points to return
 */
export const getPossibleInitialViaPositions = (params: {
  portPairsEntries: Array<
    [
      connectionName: string,
      {
        start: Omit<MHPoint, "xMoves" | "yMoves">
        end: Omit<MHPoint, "xMoves" | "yMoves">
      },
    ]
  >
  bounds: Bounds
  viaCountVariants: Array<number[]>
  reservedViaPositionsByConnectionName?: Record<string, Point[]>
}): Array<ViaPositionVariantForLinesViaCountVariant> => {
  const {
    bounds,
    portPairsEntries,
    viaCountVariants,
    reservedViaPositionsByConnectionName = {},
  } = params

  const { centroids } = getCentroidsFromInnerBoxIntersections(
    bounds,
    portPairsEntries.map(([_, portPair]) => portPair),
  )

  const result: ViaPositionVariantForLinesViaCountVariant[] = []

  for (const viaCountVariant of viaCountVariants) {
    const viaCount = viaCountVariant.reduce((acc, count) => acc + count, 0)
    const reservedViaPositionsByConnection = portPairsEntries.map(
      ([connectionName], index) =>
        (reservedViaPositionsByConnectionName[connectionName] ?? []).slice(
          0,
          viaCountVariant[index],
        ),
    )
    const reservedViaPositionKeys = new Set(
      reservedViaPositionsByConnection
        .flat()
        .map((point) => `${point.x.toFixed(6)},${point.y.toFixed(6)}`),
    )
    const reservedViaCount = reservedViaPositionsByConnection.reduce(
      (count, points) => count + points.length,
      0,
    )
    const additionalViaCountNeeded = Math.max(0, viaCount - reservedViaCount)

    let viaPositionSource: Array<{ x: number; y: number }> = centroids
      .filter(
        (point) =>
          !reservedViaPositionKeys.has(
            `${point.x.toFixed(6)},${point.y.toFixed(6)}`,
          ),
      )

    if (viaPositionSource.length < additionalViaCountNeeded) {
      // There aren't enough centroids (might not be a very hard problem)
      // we can just space the vias evenly within the node
      viaPositionSource = []
      const rows = Math.ceil(Math.sqrt(additionalViaCountNeeded))
      const cols = rows
      for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
          const candidate = {
            x:
              bounds.minX +
              ((c + 1) / (cols + 1)) * (bounds.maxX - bounds.minX),
            y:
              bounds.minY +
              ((r + 1) / (rows + 1)) * (bounds.maxY - bounds.minY),
          }
          if (
            reservedViaPositionKeys.has(
              `${candidate.x.toFixed(6)},${candidate.y.toFixed(6)}`,
            )
          ) {
            continue
          }
          viaPositionSource.push(candidate)
        }
      }
    }

    const viaPositionVariants = generateBinaryCombinations(
      additionalViaCountNeeded,
      viaPositionSource.length,
    )

    for (const viaPositionVariant of viaPositionVariants) {
      const selectedAdditionalViaPositions: Point[] = []
      for (let i = 0; i < viaPositionVariant.length; i++) {
        if (viaPositionVariant[i] === 1) {
          selectedAdditionalViaPositions.push(viaPositionSource[i])
        }
      }

      const viaPositions: Point[] = []
      let additionalViaPositionIndex = 0
      for (let i = 0; i < portPairsEntries.length; i++) {
        viaPositions.push(...reservedViaPositionsByConnection[i])
        const remainingViaCountForConnection =
          viaCountVariant[i] - reservedViaPositionsByConnection[i].length
        for (let j = 0; j < remainingViaCountForConnection; j++) {
          viaPositions.push(
            selectedAdditionalViaPositions[additionalViaPositionIndex++],
          )
        }
      }
      result.push({
        viaPositions,
        viaCountVariant,
      })
    }
  }

  return result
}

File: rust/intra-node-routing/src/multi_head_poly_line_intra_node_solver/get_possible_initial_via_positions.rs
use super::{compute_via_count_variants::PortPairsEntries, generate_binary_combinations::generate_binary_combinations,
    get_centroids_from_inner_box_intersections::{Segment,get_centroids_from_inner_box_intersections}};
use crate::specialized_utils::{math::{Point,SpecializedMath},get_bounds_from_node_with_port_points::Bounds};

pub struct ViaPositionVariant { pub via_positions: Vec<Point>, pub via_count_variant: Vec<usize> }

pub fn get_possible_initial_via_positions(pairs: &PortPairsEntries, bounds: &Bounds, variants: &[Vec<usize>], math: SpecializedMath) -> Result<Vec<ViaPositionVariant>, String> {
    let segments: Vec<_> = pairs.iter().map(|(_, (a,b))| Segment { start:Point{x:a.x,y:a.y},end:Point{x:b.x,y:b.y},connection_name:None }).collect();
    let centroids = get_centroids_from_inner_box_intersections(bounds,&segments,math).centroids;
    let mut result = Vec::new();
    for variant in variants {
        let count = variant.iter().sum::<usize>();
        let mut source = centroids.clone();
        if centroids.len() < count {
            source.clear();
            let rows = (count as f64).sqrt().ceil() as usize;
            let cols = rows;
            for r in 0..rows { for c in 0..cols {
                source.push(Point { x:bounds.min_x+(c+1) as f64/(cols+1) as f64*(bounds.max_x-bounds.min_x),y:bounds.min_y+(r+1) as f64/(rows+1) as f64*(bounds.max_y-bounds.min_y) });
            }}
        }
        for position_variant in generate_binary_combinations(count,source.len())? {
            let mut positions = Vec::new();
            for (index, &include) in position_variant.iter().enumerate() { if include==1 { positions.push(source[index]); } }
            result.push(ViaPositionVariant {via_positions:positions,via_count_variant:variant.clone()});
        }
    }
    Ok(result)
}
